include/deal.II/non_matching/closest_surface_point.h

```diff
@@ -0,0 +1,164 @@
+// ------------------------------------------------------------------------
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2025 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// Part of the source code is dual licensed under Apache-2.0 WITH
+// LLVM-exception OR LGPL-2.1-or-later. Detailed license information
+// governing the source code and code contributions can be found in
+// LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
+//
+// ------------------------------------------------------------------------
+
+#ifndef dealii_closest_surface_point
+#define dealii_closest_surface_point
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/quadrature.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/read_vector.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+
+
+  /**
+   * @brief A class for computing the closest points on a surface defined by a level set function.
+   *
+   * This class implements algorithms to find the closest points on a surface
+   * (zero level set) to a given set of query points. The surface is implicitly
+   * defined by a level set function discretized on a finite element mesh. The
+   * class uses Newton's method to iteratively find the closest surface points
+   * by minimizing the distance function subject to the constraint that the
+   * point lies on the zero level set.
+   *
+   * - Level set functions discretized on DoFHandler with cartesian grid.
+   * - Configurable Newton iteration parameters (tolerance, maximum iterations)
+   * - Supports active cells and MG cells
+   *
+   * @tparam dim The spatial dimension of the problem
+   * @tparam Number The scalar type of vector used to store the level set function values
+   *
+   * @note The level set function should be negative inside the domain, positive outside,
+   *       and zero on the surface boundary.
+   *
+   * @warning The Newton iteration may not converge for points that are too far from the surface
+   *          or in regions where the level set function has poor conditioning.
+   */
+  template <int dim, class Number>
+  class ClosestSurfacePoint : public EnableObserverPointer
```

Comment:
@mwichro this algorithm looks very interesting to us. We are currently working on closest point projection in the context of extrapolating quantities from an implicitly given interface and for geometric reinitialization in level-set frameworks. 

What I’m not entirely clear on from the description and the algorithm is whether your method can find the closest surface points for any cell in the domain (potentially a non-ghost cell that is owned by a different process in a p::d::triangulation), or if it only works for cells that are directly adjacent to the interface and are either ghost cells or locally owned ones?